### PR TITLE
Issue #904 Fix scope of allowedToProxy flag.

### PR DESCRIPTION
**Pull request by serac**
A new field, proxiedBy, is introduced on TicketGrantingTicket to track the service that generates a proxy-granting ticket to facilitate looking up the corresponding RegisteredService at the time proxy tickets are issued. The allowedToProxy flag on the proxiedBy service determines whether or not the proxy ticket request can proceed.

### CODE REVIEW

**Comment by serac** [comment]
Note that, unfortunately, this pull introduces an API change, which triggered the minor version bump and the target branch, 3.6.x.

**Comment by mmoayyed** [comment]
Thank you. I'll test this out tomorrow.

What's our plan to get this to 4.x? assuming change goes through? Do you want me to handle the merge there as well?

**Comment by mmoayyed** [comment]
This works fine. I'd ask @leleuj  or @battags to also review.

**Comment by leleuj** [comment]
The `allowedToProxy` flag has always applied to the proxying service and proxied service. With this change, the check would only apply on the proxying service and no more on the proxied service. In that case, why is the check in the `delegateTicketGrantingTicket` method (when the PGT is created) not sufficient? Why do we need a check in the `grantServiceTicket` method?

**Comment by mmoayyed** [comment]
Because when you are trying to create a PT, you still want to make sure the PGT is authorized to proxy, because the configuration may have changed for the service since the that PGT was created. 

For example, when service is allowed to proxy. we'll create a PGT and subsequent PTs off of that.
Then, the configuration changes to disallow a service from proxying (Support the underlying registered service data changes). Then, after this change when PT requests come in, we still need to check PGT's ability to proxy.

**Comment by serac** [comment]
@leleuj, `allowedToProxy` should _not_ be applied to the proxied service because I do not want to grant any special privilege to that service. In terms of security policy, I'm delegating authentication to the proxying service by ticking a box on the service manager. That's an increased security privilege and one we grant on a case-by-case basis after security review. I do not want to have to check that box for the proxied service because that would grant privileges I explicitly do not want to grant. That is the exact use case I'm trying to fix with this pull request.

**Comment by leleuj** [comment]
@mmoayyed: I understand the use case, but it's rare one, isn't it? As we could just wait for the PGT to expire, couldn't we.

@serac: so I don't think it's a bug, it's a change of meaning: before, "allowedToProxy" meant "allowed to create PGT" OR "allowed to create PT", now it will only mean that it allows PGT creation. A service won't be able to refuse to be used through a proxy, right?

**Comment by serac** [comment]
@leleuj, it's a change in meaning that brings the functionality back in line with the documentation of the flag. From https://wiki.jasig.org/display/CASUM/Adding+a+New+Service:

> Allowed to proxy - whether the service can retrieve Proxy Granting Tickets; defaults to true.
> If set to false, CAS will immediately stop vending Proxy Granting Tickets to this services.
> Existing PGTs will still be honored until they expire.

The flag thus impacts checks on PGTs exclusively, and the current implementation is broken in that respect. This patch makes the behavior consistent with the documentation, and in my opinion, the expectations of deployers.

File: cas-server-core/src/main/java/org/jasig/cas/ticket/TicketGrantingTicketImpl.java [hunk 1]
```diff
@@ -77,19 +82,24 @@ public TicketGrantingTicketImpl() {
...
      * @param id the id of the Ticket
+     * @param proxiedBy Service that produced this proxy ticket.
      * @param ticketGrantingTicket the parent ticket
      * @param authentication the Authentication request for this ticket
      * @param policy the expiration policy for this ticket.
      * @throws IllegalArgumentException if the Authentication object is null
      */
     public TicketGrantingTicketImpl(final String id,
-        final TicketGrantingTicketImpl ticketGrantingTicket,
+        final String proxiedBy, final TicketGrantingTicketImpl ticketGrantingTicket,
```
Comment (leleuj) [inline cas-server-core/src/main/java/org/jasig/cas/ticket/TicketGrantingTicketImpl.java]: BTW, shouldn't we rename `ticketGrantingTicket` as `parentTGT` or something to make that obvious?

File: cas-server-core/src/main/java/org/jasig/cas/ticket/TicketGrantingTicketImpl.java [hunk 2]
```diff
@@ -68,7 +68,12 @@
     @Lob
     @Column(name="SERVICES_GRANTED_ACCESS_TO", nullable=false)
     private final HashMap<String,Service> services = new HashMap<String, Service>();
-    
+
+    /** Service that produced a proxy-granting ticket. */
+    @Column(name="PROXIED_BY", nullable=true)
+    private String proxiedBy;
```
Comment (leleuj) [inline cas-server-core/src/main/java/org/jasig/cas/ticket/TicketGrantingTicketImpl.java]: I guess it should be `final`.

**Comment by leleuj** [comment]
OK. I'm in line. So we don't add a new check to allow or not a service to be used by a proxy, do we.

+1 with a few comments

File: cas-server-core/src/main/java/org/jasig/cas/ticket/TicketGrantingTicketImpl.java (hunk 1 above)
Comment (serac) [inline cas-server-core/src/main/java/org/jasig/cas/ticket/TicketGrantingTicketImpl.java]: Agreed that would be clearer. Will change.

**Comment by serac** [comment]
I will formulate a separate pull request against master once this is accepted.

**Comment by mmoayyed** [comment]
Thanks for keeping with this. 

Just one small note, that this case is not uncommon at all. Consider this scenario:

1. An admin allows application X to proxy and redeploys CAS
2. Application X happily proxies for 1 year
3. Admin decides that Application X should not be allowed to proxy anymore.
4. Admin goes to the management webapp, or the JSON file and turns that check off, without restarting CAS.
5. CAS loads the change.

Now, if that change in the registry is not evaluated when granting PTsX, will continue to receive proxy tickets, until the CAS server is restarted, which is definitely a bug, or at least, a very frustrating surprise.

File: cas-server-core/src/main/java/org/jasig/cas/ticket/TicketGrantingTicketImpl.java (hunk 2 above)
Comment (serac) [inline cas-server-core/src/main/java/org/jasig/cas/ticket/TicketGrantingTicketImpl.java]: It's not set in all cases -- there's a no-arg constructor. Also, it's consistent with other non-collection fields.

**Comment by mmoayyed** [comment]
So it looks like we have a few suggestions here to account for. Once applied, I'll merge.

**Comment by leleuj** [comment]
Just trying to get a better understanding: after the proxy checkbox has been turned off, proxy tickets will be created till the PGT is not expired, let's a say, a few hours, right?

**Comment by mmoayyed** [comment]
sure. PGTs share the same expiration policy as TGT. Default lasts for 8 hours, while it could be for a full week :)

**Comment by mmoayyed** [comment]
To clarify, with this pull proxy tickets should immediately be blocked if the checkbox is turned off. Without this pull, they will continue to be issued, which is relevant the bug addressed here. While this pull fixes the proxy configuration, it also sort of explains why the check is necessary when granting PTs in `grantServiceTicket`

**Comment by mmoayyed** [comment]
@serac I have created a milestone to track the 3.6 release line. You could resume working on separate pulls for 4.x and master. Thanks :)

**Comment by serac** [comment]
@mmoayyed , I was planning for only a patch against master since the target fix version is 4.1 in any case due to the API change. Make sense?

**Comment by mmoayyed** [comment]
Ah, sure. You're right.

**Pull request merged**